File: #98_Validate_Binary_Search_Tree/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public boolean isValidBST(TreeNode root) {
    	if (root == null) return true;
    	List<Integer> inorder = new ArrayList<Integer>();
    	inorderTraversal(root, inorder);
    	int len = inorder.size();
    	if (len == 1) return true;
    	for (int i = 1; i < len; i++) {
    		if (inorder.get(i - 1) >= inorder.get(i)) return false;
    	}
    	return true;
    }

    private static void inorderTraversal(TreeNode node, List<Integer> list) {
    	if (node == null) return;
    	inorderTraversal(node.left, list);
    	list.add(node.val);
    	inorderTraversal(node.right, list);
    }
}

File: #98_Validate_Binary_Search_Tree/Solution.py
'''
in-order traversal
'''

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.prev = None

	# @param {TreeNode} root
	# @return {boolean}
	def isValidBST(self, root):
		self.prev = None
		cur = root
		return self.inOrder(cur)

	def inOrder(self, root):
		if not root:
			return True
		if root.left:
			if not self.inOrder(root.left):
				return False
		if self.prev and root.val <= self.prev.val:
			return False
		self.prev = root# set prev for next level
		if root.right:
			if not self.inOrder(root.right):
				return False
		return True

sol = Solution()
root = TreeNode(1)
n1 = TreeNode(1)
# n2 = TreeNode(3)
root.left = n1
# root.right = n2
print sol.isValidBST(root)